include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+    return generateCommands(random, size);
+  }
+
+private:
+  struct CommandEntry {
+    CommandEntry(Random &&aRandom, Shrinkable<CmdSP> &&aShrinkable)
+        : random(std::move(aRandom))
+        , shrinkable(std::move(aShrinkable)) {}
+
+    Random random;
+    Shrinkable<CmdSP> shrinkable;
+  };
+
+  struct CommandSequence {
+    CommandSequence(const GenFunc &func, int sz)
+        : genFunc(func)
+        , size(sz) {}
+
+    Model initialState;
+    GenFunc genFunc;
+    int size;
+    std::vector<CommandEntry> entries;
+
+
+    void repairEntriesFrom(std::size_t start) {
+      for (auto i = start; i < entries.size(); i++) {
+        if (!repairEntryAt(i)) {
+          entries.erase(begin(entries) + i--);
+        }
+      }
+    }
+
+    bool repairEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        const auto cmd = entry.shrinkable.value();
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+
+        return regenerateEntryAt(i);
+      }
+
+      return true;
+    }
+
+    bool regenerateEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        entry.shrinkable = genFunc()(entry.random, size);
+        const auto cmd = entry.shrinkable.value();
+        return true;
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+      } catch (const GenerationFailure &failure) {
+        // Just return false below
+      }
+
+      return false;
+    }
+  };
+
+  struct ParallelCommandSequence {
+    ParallelCommandSequence(CommandSequence prefix,
+                            CommandSequence left,
+                            CommandSequence right)
+        : prefix(prefix)
+        , left(left)
+        , right(right) {}
+
+    CommandSequence prefix;
+    CommandSequence left;
+    CommandSequence right;
+  };
+
+  Shrinkable<ParallelCommands<Cmd>> generateCommands(const Random &random,
+                                                     int size) const {
+    return shrinkable::map(generateParallelSequence(random, size),
+                           [](const ParallelCommandSequence &sequence) {
+                             ParallelCommands<Cmd> cmds;
+                             copyCommandEntries(sequence.prefix, cmds.prefix);
+                             copyCommandEntries(sequence.left, cmds.left);
+                             copyCommandEntries(sequence.right, cmds.right);
+                             return cmds;
+                           });
+  }
+
+  template <typename T>
+  static void copyCommandEntries(const CommandSequence &source,
+                                 std::vector<T> &dest) {
+    const auto &entries = source.entries;
+    dest.reserve(entries.size());
+    std::transform(
```

Comment:
Can ShrinkValueIterator be uesd? entries is a vector<CommandEntry> (which has a member of type Shrinkable<CmdSp>) not a vector<Shrinkable<CmdSp>>.